lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1494,22 +1494,31 @@ class HSAQueue final : public KalmarQueue
     bool isEmpty() override {
         // Have to walk asyncOps since it can contain null pointers (if event is waited on and removed)
         // Also not all commands contain signals.
-        for (int i = 0; i < asyncOps.size(); ++i) {
-            if (asyncOps[i] != nullptr) {
-                auto &asyncOp = asyncOps[i];
+        bool oldestNoSignal=false;
+        for (int i = asyncOps.size()-1; i >= 0;  i--) {
+            auto &asyncOp = asyncOps[i];
+            if (asyncOp != nullptr) {
                 hsa_signal_t signal = *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
                 if (signal.handle) {
+                    oldestNoSignal=false;
                     hsa_signal_value_t v = hsa_signal_load_relaxed(signal);
```

Comment:
Are we sure relaxed with no fence is what we want here? I believe that this is concurrently written by the CP, is it not?

Reply:
This is unchanged from code before, but probably should be changed.  It comes down to what effect is implied on the memory model if a user submits a command and then checks if the associated accelerator view is empty - probably would assume that the results of that command are visible.